File: src/packages/agent_core/builtin_tools/web_fetch/provider.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True, slots=True)
class WebFetchResult:
    url: str
    content: str
    title: str
    truncated: bool

    def to_json(self) -> dict[str, object]:
        return {
            "content": self.content,
            "title": self.title,
            "url": self.url,
            "truncated": self.truncated,
        }


class WebFetchProvider(Protocol):
    async def fetch(self, *, url: str, max_length: int) -> WebFetchResult: ...


__all__ = ["WebFetchProvider", "WebFetchResult"]

File: src/services/worker/internal/tools/builtin/web_fetch/provider.go
package webfetch

import "context"

type Result struct {
	URL       string
	Title     string
	Content   string
	Truncated bool
}

func (r Result) ToJSON() map[string]any {
	payload := map[string]any{
		"url":       r.URL,
		"title":     r.Title,
		"content":   r.Content,
		"truncated": r.Truncated,
	}
	return payload
}

type Provider interface {
	Fetch(ctx context.Context, url string, maxLength int) (Result, error)
}

type HttpError struct {
	StatusCode int
}

func (e HttpError) Error() string {
	return "http error"
}
